src/software/ai/navigator/trajectory/trajectory_planner.cpp

```diff
@@ -0,0 +1,321 @@
+    // Avoid finding the first non-collision time if the cache sub-trajectory
+    // has a collision time.
+    if (sub_traj_with_cost.has_value() &&
+        sub_traj_with_cost->collision_duration_front_s < sub_traj_duration_s)
+    {
+        first_non_collision_time = sub_traj_with_cost->collision_duration_front_s;
+    }
+    else
+    {
+        first_non_collision_time = getFirstNonCollisionTime(
+            trajectory, possible_collisions_indices, obstacles, search_end_time_s);
+    }
+    traj_with_cost.collision_duration_front_s = first_non_collision_time;
+
+    /**
+     * Find the duration we're within an obstacle before search_end_time_s
+     */
+    double last_non_collision_time = getLastNonCollisionTime(
+        trajectory, possible_collisions_indices, obstacles, search_end_time_s);
+    traj_with_cost.collision_duration_back_s =
+        search_end_time_s - last_non_collision_time;
+
+    /**
+     * Get the first collision time, excluding the time at the start and end of path
+     * that we may be in an obstacle for.
+     */
+    if (sub_traj_with_cost.has_value() &&
+        sub_traj_with_cost->first_collision_time_s < sub_traj_duration_s)
+    {
+        traj_with_cost.first_collision_time_s =
+            sub_traj_with_cost->first_collision_time_s;
+        traj_with_cost.colliding_obstacle = sub_traj_with_cost->colliding_obstacle;
+    }
+    else
+    {
+        std::pair<double, ObstaclePtr> collision =
+            getFirstCollisionTime(trajectory, possible_collisions_indices, obstacles,
+                                  first_non_collision_time, last_non_collision_time);
+        traj_with_cost.first_collision_time_s = collision.first;
+        traj_with_cost.colliding_obstacle     = collision.second;
+    }
+
+    traj_with_cost.cost = calculateCost(traj_with_cost);
+
+    return traj_with_cost;
+}
+
+double TrajectoryPlanner::calculateCost(
+    const TrajectoryPathWithCost &traj_with_cost) const
+{
+    double total_cost = traj_with_cost.traj_path.getTotalTime();
+
+    // Add a large cost if the trajectory collides with an obstacle
+    // Note that this ignores collisions that may be in at the
+    // start of the trajectory as those are unavoidable by all trajectories.
+    if (traj_with_cost.colliding_obstacle != nullptr)
+    {
+        total_cost += 6.0;
+    }
+
+    // The closer the collision is to the destination, the lower its cost will be
+    Point first_collision_position =
+        traj_with_cost.traj_path.getPosition(traj_with_cost.first_collision_time_s);
+    Point destination = traj_with_cost.traj_path.getDestination();
+    total_cost += (first_collision_position - destination).length();
+
+    total_cost += std::max(
+        0.0, (MAX_FUTURE_COLLISION_CHECK_SEC - traj_with_cost.first_collision_time_s));
+
+    total_cost += 3 * traj_with_cost.collision_duration_front_s;
+
+    total_cost += 1 * traj_with_cost.collision_duration_back_s;
+
+    return total_cost;
```

Comment:
this may be something we want to spend some effort tuning -> this could be a good task for someone on the motion planning project

Reply:
is it worth having it configurable via dynamic params? It might be too specific to be useful as a dynamic param

Reply:
I think the optimization of trajectories is still an open ended discussion, as there are possible alternative approaches that don't use cost at all. So not sure if the effort of tuning this is needed. Though it's a good idea to keep an eye out and see if it sometimes picks weird/unexpected trajectories, then we can look at those specific scenarios.